### PR TITLE
Add DRG based VCN peering

### LINKED ISSUE #14: Provide ability to peer Management and Workload Cluster VCN
**What would you like to be added**:
Users should have the ability to peer management cluster VCN with workload cluster VCN. There are various techniques that can be used such as Local Peering Gateway or Dynamic Routing Gateway. Using Dynamic Routing Gateway will provide a single approach to peer VCNs in same as well as different regions and hence recommended as an initial implementation.  

**Why is this needed**:
This is required for private cluster support.

### PR DESCRIPTION
**Pull request by shyamradhakrishnan**
**What this PR does / why we need it**:
This PR add DRG based VCN peering to CAPOCI

The CCM tests have been refactored to use latest image and remove the extra image which was required when CCM was not supporting 1.22

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #14 

## Functionality
Users can peer workload cluster VCN with any other VCN(typically management cluster VCN) via 2 methods. If both management cluster and workload cluster are in same region, then a single DRG can be used, and both the VCNs can be attached to the same VCN via VCN Attachment. If the management cluster and workload cluster are in separate regions, then they have to be peered using RPC.

### Case 1: Same region
```yaml
  networkSpec:
    vcnPeering:
      drg:
        id:
      peerRouteRules:
        - vcnCIDRRange: "10.0.0.0/16"
```

In this case, user provides the above YAML. In the DRG, CAPOCI will create a VCN Attachment with workload cluster VCN. DRG will not be managed in this case, thew ID has to be provided. The DRG can be shared across multiple workload clusters.

In the workload cluster VCN, the route rules specified will be added as DRG as the destination so that traffic to the management cluster will go via DRG.

### Case 2: Different region

```yaml
  networkSpec:
    vcnPeering:
      drg:
        manage: true
      peerRouteRules:
        - vcnCIDRRange: "10.0.0.0/16"
      remotePeeringConnections:
        - managePeerRPC: true
          peerDRGId: 
          peerRegionName: "us-phoenix-1"
```

In this case, there will be a DRG per workload cluster. The DRG will be created by CAPOCI. After the DRG is created, CAPOCI will create RPCs in both the local and peer DRG. And then a connection will be initiated between the 2 RPCs.

### CODE REVIEW

**Comment by shyamradhakrishnan** [comment]
Unit test
```
?   	github.com/oracle/cluster-api-provider-oci	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/api/v1beta1	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/config	[no test files]
ok  	github.com/oracle/cluster-api-provider-oci/cloud/ociutil	0.478s	coverage: 30.3% of statements
ok  	github.com/oracle/cluster-api-provider-oci/cloud/scope	176.808s	coverage: 81.9% of statements
?   	github.com/oracle/cluster-api-provider-oci/cloud/scope/mocks	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/compute	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/compute/mock_compute	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/identity	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/identity/mock_identity	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/networkloadbalancer	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/networkloadbalancer/mock_nlb	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/vcn	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/vcn/mock_vcn	[no test files]
ok  	github.com/oracle/cluster-api-provider-oci/controllers	26.764s	coverage: 70.3% of statements
```

File: templates/cluster-template-local-vcn-peering.yaml [hunk 1]
```diff
@@ -0,0 +1,448 @@
...
+            cloud-provider: external
+            provider-id: oci://{{ ds["id"] }}
+---
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: MachineDeployment
+metadata:
+  name: "${CLUSTER_NAME}-md-0"
+spec:
+  clusterName: "${CLUSTER_NAME}"
+  replicas: ${WORKER_MACHINE_COUNT}
```
Comment (joekr) [inline templates/cluster-template-local-vcn-peering.yaml:433]: Should this be `NODE_MACHINE_COUNT`?

**Review by joekr — CHANGES_REQUESTED** [review]

File: cloud/scope/cluster.go [hunk 2]
```diff
@@ -54,6 +54,7 @@ type ClusterScopeParams struct {
 	IdentityClient        identityClent.Client
 	Region                string
 	OCIAuthConfigProvider common.ConfigurationProvider
+	ClientProvider        *ClientProvider
```
[the remark this answers is not in the document]
Comment (joekr) [reply cloud/scope/cluster.go:57]: I think I understand the way. `ClientProvider` is used for the peering region while the `VCNClient` is used for the "current" region

[the remark this answers is not in the document]
Comment (shyamradhakrishnan) [reply cloud/scope/cluster.go:58]: yes, I did not want to touch any code as that would be slightly risky and will need to change a whole lot of unit tests. WE need to use the clients like Normal for the Clusters region and only use the ClientProvider in special cases like the VCN peering thing. Makes sense?

File: cloud/scope/drg_reconciler.go [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+		return nil, err
+	}
+	return &response.Drg, nil
+}
+
+func (s *ClusterScope) GetDRGName() string {
+	if s.OCICluster.Spec.NetworkSpec.VCNPeering.DRG.Name != "" {
+		return s.OCICluster.Spec.NetworkSpec.VCNPeering.DRG.Name
+	}
+	return fmt.Sprintf("%s", s.OCICluster.Name)
```
[the remark this answers is not in the document]
Comment (shyamradhakrishnan) [reply cloud/scope/drg_reconciler.go:152]: The general consensus was that we should not add -<type> for anything, for example our VCN name, subnet name etc dont have the type appended.

File: cloud/scope/drg_rpc_attachment_reconciler.go [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+					PeerRegionName: common.String(rpcSpec.PeerRegionName),
+				},
+			})
+			if err != nil {
+				return err
+			}
+			s.Logger.Info("Connect request initiated for local and peer RPCs", "rpcId", rpcSpec.RPCConnectionId)
+		}
+		if localRpc.PeeringStatus != core.RemotePeeringConnectionPeeringStatusPeered {
+			err := wait.PollWithContext(ctx, PollInterval, RequestTimeout, func(ctx context.Context) (done bool, err error) {
```
[the remark this answers is not in the document]
Comment (shyamradhakrishnan) [reply cloud/scope/drg_rpc_attachment_reconciler.go:132]: This is different from RPC provisioning, this is connection and the API we use as well as the states are different. Does that make sense?

File: templates/cluster-template-local-vcn-peering.yaml (hunk 1 above)
Comment (shyamradhakrishnan) [reply templates/cluster-template-local-vcn-peering.yaml]: Done

File: controllers/ocicluster_controller_test.go [hunk 5]
```diff
@@ -104,7 +104,7 @@ func TestOCIClusterReconciler_reconcile(t *testing.T) {
 			Spec:       infrastructurev1beta1.OCIClusterSpec{},
 			Status:     infrastructurev1beta1.OCIClusterStatus{},
 		}
-		recorder = record.NewFakeRecorder(10)
+		recorder = record.NewFakeRecorder(20)
```
[the remark this answers is not in the document]
Comment (shyamradhakrishnan) [reply controllers/ocicluster_controller_test.go:107]: This is the number of events which will be stored in the fake recorder.

File: cloud/scope/drg_reconciler.go [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+	"github.com/pkg/errors"
+)
+
+// ReconcileDRG tries to move the DRG to the desired OCICluster Spec
+func (s *ClusterScope) ReconcileDRG(ctx context.Context) error {
+	if !s.isPeeringEnabled() {
+		s.Logger.Info("VCN Peering is not enabled, ignoring reconciliation")
+		return nil
+	}
+	if s.OCICluster.Spec.NetworkSpec.VCNPeering.DRG == nil {
```
[the remark this answers is not in the document]
Comment (shyamradhakrishnan) [reply cloud/scope/drg_reconciler.go]: Done, please check if the change what I have done makes sense

File: cloud/scope/drg_rpc_attachment_reconciler.go [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+		if err != nil {
+			if ociutil.IsNotFound(err) {
+				return true, nil
+			} else {
+				return true, err
+			}
+		}
+		if rpc == nil || rpc.LifecycleState == core.RemotePeeringConnectionLifecycleStateTerminated {
+			return true, nil
+		}
```
[the remark this answers is not in the document]
Comment (shyamradhakrishnan) [reply cloud/scope/drg_rpc_attachment_reconciler.go:299]: The states are available, terminating, terminated and available. I think we should wait for it to reach terminated, it it does not reach the loop will timeout and the reconciliation will error out. Dees that sound ok?

File: cloud/scope/drg_vcn_attachment_reconciler.go [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	if len(attachments.Items) == 0 {
+		return nil, nil
+	} else if len(attachments.Items) > 1 {
+		return nil, errors.New("found more than one DRG VCN attachment to same VCN")
```
[the remark this answers is not in the document]
Comment (shyamradhakrishnan) [reply cloud/scope/drg_vcn_attachment_reconciler.go]: I have added more help in the error, is that ok? This is one of the corner cases where the api returns a list and 99% of times, users wont touch anything but we still need to handle the code.

File: cloud/scope/drg_reconciler.go [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+	s.getDRG().ID = drg.Id
+	s.Logger.Info("Successfully created DRG", "drg", *drg.Id)
+	return nil
+}
+
+// GetDRG retrieves the Cluster's core.Drg using the one of the following methods
+//
+// 1. the OCICluster's spec Drg
+//
+// 2. Listing the Drgs for the Compartment (by ID) and filtering by tag
```
Comment (joekr) [inline cloud/scope/drg_reconciler.go:72]: I love this kind of method comment and how it explains the high level steps! 😻

**Review by joekr — DISMISSED** [review]

**Review by joekr — APPROVED** [review]

**Pull request merged**